File: MMORPGServer/Proto/CreateLuaProto/RoleOperation_LogOnGameServerProto.lua
--客户端发送登录区服消息
RoleOperation_LogOnGameServerProto = { ProtoCode = 10001, AccountId = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleOperation_LogOnGameServerProto.__index = RoleOperation_LogOnGameServerProto;

function RoleOperation_LogOnGameServerProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleOperation_LogOnGameServerProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleOperation_LogOnGameServerProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.AccountId);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleOperation_LogOnGameServerProto.GetProto(buffer)

    local proto = RoleOperation_LogOnGameServerProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.AccountId = ms:ReadInt();

    ms:Dispose();
    return proto;
end

File: MMORPGServer/GameServerApp/GameServerApp/Proto/Role/RoleOperation_LogOnGameServerProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2020-03-15 10:58:27
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 客户端发送登录区服消息
/// </summary>
public struct RoleOperation_LogOnGameServerProto : IProto
{
    public ushort ProtoCode { get { return 10001; } }

    public int AccountId; //账户ID

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(AccountId);
            return ms.ToArray();
        }
    }

    public static RoleOperation_LogOnGameServerProto GetProto(byte[] buffer)
    {
        RoleOperation_LogOnGameServerProto proto = new RoleOperation_LogOnGameServerProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.AccountId = ms.ReadInt();
        }
        return proto;
    }
}